query-service-impl/src/main/java/org/hypertrace/core/query/service/postgres/converters/PostgresFunctionConverter.java

```diff
@@ -0,0 +1,280 @@
+    if (index == -1) {
+      throw new IllegalArgumentException("Unable to parse period : " + period);
+    }
+    int periodValue = Integer.parseInt(period.substring(0, index));
+    String timeUnit = period.substring(index + 1);
+    switch (timeUnit.toUpperCase()) {
+      case "SECONDS":
+        return TimeUnit.SECONDS.toMillis(periodValue);
+      case "MINUTES":
+        return TimeUnit.MINUTES.toMillis(periodValue);
+      case "HOURS":
+        return TimeUnit.HOURS.toMillis(periodValue);
+      default:
+        throw new UnsupportedOperationException("Unsupported time unit : " + timeUnit);
+    }
+  }
+
+  private String functionToString(
+      Function function, java.util.function.Function<Expression, String> argumentConverter) {
+    String argumentString =
+        function.getArgumentsList().stream()
+            .map(argumentConverter)
+            .collect(Collectors.joining(","));
+
+    return function.getFunctionName() + "(" + argumentString + ")";
+  }
+
+  private String functionToStringForAvgRate(
+      Function function,
+      java.util.function.Function<Expression, String> argumentConverter,
+      ExecutionContext executionContext) {
+
+    String columnName = argumentConverter.apply(function.getArgumentsList().get(0));
+    String rateIntervalInIso =
+        function.getArgumentsList().size() == 2
+            ? function.getArgumentsList().get(1).getLiteral().getValue().getString()
+            : DEFAULT_AVG_RATE_SIZE;
+    long rateIntervalInSeconds = isoDurationToSeconds(rateIntervalInIso);
+    long aggregateIntervalInSeconds =
+        executionContext
+            .getTimeSeriesPeriod()
+            .or(executionContext::getTimeRangeDuration)
+            .orElseThrow()
+            .getSeconds();
+
+    return String.format(
+        "SUM(%s) / %s", columnName, (double) aggregateIntervalInSeconds / rateIntervalInSeconds);
+  }
+
+  private String convertCount() {
+    if (tableDefinition.getCountColumnName() != null) {
+      return SUM_FUNCTION + "(" + tableDefinition.getCountColumnName() + ")";
+    } else {
+      return "COUNT(*)";
+    }
+  }
+
+  private String toPostgresPercentile(
+      Function function, java.util.function.Function<Expression, String> argumentConverter) {
+    int percentileValue =
+        this.getPercentileValueFromFunction(function)
+            .orElseThrow(
+                () ->
+                    new UnsupportedOperationException(
+                        String.format(
+                            "%s must include an integer convertible value as its first argument. Got: %s",
+                            QUERY_FUNCTION_PERCENTILE, function.getArguments(0))));
+    Expression expr =
+        function.getArgumentsList().stream()
+            .filter(Expression::hasAttributeExpression)
+            .findFirst()
+            .orElseThrow(
+                () -> new IllegalArgumentException("Attribute expression not found : " + function));
+    boolean isTdigest =
```

Comment:
should we avoid this client change? If we go with this path, we need to do a similar for distinct_count_hll, and all.
If, we do this, we should do this across - pinot + prom + postgres. For now, can we handle this internally?

Reply:
Fixed